### PR TITLE
Update CI and project properties

### LINKED ISSUE #254: CI build fails because of `.netcoreapp3.1` TFM
Probably shouldn't have merged #252 and #253 without further checks. Shouldn't be too difficult to fix, but let's do so before merging anything else.

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #254 

Changes:
- dependabot tracks GitHub actions
- CI works entirely on GitHub actions, cmd/sh scripts removed
- SourceLink added
- Description for tool package added
- net7 TFM for tool added
- Tests run for netcoreapp3.1, net6 and net7

~::WARNING:: need to add `NUGET_AUTH_TOKEN` into repository or org secrets to publish packahes to nuget.org~

### CODE REVIEW

**Comment by stakx** [comment]
@sungam3r, apologies again for letting this sit around for so long. I haven't reviewed this one yet, but if you'd still like to make some changes to the CI configuration, could you please rebase this on top of current `master`?

**Comment by sungam3r** [comment]
@stakx Done. Let's discuss `Publish Nuget packages to Nuget registry`  step. How did you publish to Nuget.org before? I can remove this step at all. It should not be run on CI (at least preview packages should be published at GitHub package registry). Or I can add other workflow like this [one](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/.github/workflows/publish-code.yml).

**Comment by stakx** [comment]
> How did you publish to Nuget.org before? I can remove this step at all. It should not be run on CI[.]

I have only published a single release so far, and I had to do that manually... which isn't ideal. I completely agree with you that publishing should happen automatically by CI, whenever a version tag is pushed.

**Comment by sungam3r** [comment]
Different strategies exist. I see you use MinVer here. This means automatic version control if I understand correctly. I'm not familiar with MinVer but I know some repos that use it. Personally I prefer semi-automatic (or semi-manual if you like) versioning. Just look at workflows here https://github.com/graphql-dotnet/parser/tree/master/.github/workflows to make an idea. If you like it then I can switch and delete MinVer reference. Otherwise, we need to understand how to integrate MinVer features into CI/publishing nugets.

**Comment by stakx** [comment]
@sungam3r, I haven't studied the CI setup in detail yet. I'll have to give it a closer look. TBH I may not even have a clear preference, as long as the release process it simple. Ideally we could just create a `vX.Y.Z` tag (either by manually pushing it to the repo, or by creating a release on the GitHub webpage) and CI would then release version X.Y.Z for us.

Let me take a closer look at MinVer and the thingy you linked to above. And perhaps in the meantime let me know if you have a specific rexommendation for a release workflow.

**Comment by stakx** [comment]
Alright, I've taken that look.

What I like about MinVer is the way how it takes care about some things like setting `[AssemblyVersion]` etc. I wouldn't want to do that manually.

I also like how you trigger the NuGet publishing process in your `graphql-dotnet/parser` example using `on.release.types: [ 'published' ]`. (If we keep MinVer around, which I think is my current preference, it might make more sense to change the trigger to one that also requires a version tag.)

One thing I think isn't strictly necessary is uploading the `.nupkg` as a GitHub release artifact. They're already on NuGet, and it's much easier to install them from there than downloading a package and putting it in some package source folder. Or am I missing something?

So if we take those things together, why don't we:

* Do releases by creating a release (including a tag in the Git repo) on GitHub;
* which triggers a GitHub Actions pipeline that builds, tests, and pushes the NuGet package;
* and we keep MinVer to put the right version numbers in the assembly and NuGet manifest.

That being said, I'm open to suggestions. You've maybe done this more often than I have.

(Why aren't you a maintainer at this repo btw.? Given your long-standing involvement and interest in this project, that would seem to make sense. :wink:)

**Comment by sungam3r** [comment]
>  They're already on NuGet ... Or am I missing something?

I think no. Someone writes PR. CI triggers, builds code, runs tests. No nuget publishing here, or you will end up with hundreds trash versions of package on nuget.org. OK, then after review and ~20 new commits you merge this PR in `master` - "stable" branch so you **may** have desire to publish package. Where? Either nuget.org or some other artifacts repo (like buffer for work in progres). Really, it is up to you - publish to nuget.org **each** commit to `master` or to do it manually only sometimes accumulating changes from 2/3/11/45 commits. In graphql-dotnet org repos we chose the second strategy for publishing to nuget.org but **each** commit to `master` or `develop` still results in publishing packages to nuget.pkg.github.com . Why? Because someone may have a strong need to obtain code-fixes (or test features) from these "internal" releases before "official external" release appears on GitHub/nuget.org. It is possible that for the first strategy MinVer will be a suitable decision - I do not know.

>  Do releases by creating a release (including a tag in the Git repo) on GitHub;
    which triggers a GitHub Actions pipeline that builds, tests, and pushes the NuGet package;
    and we keep MinVer to put the right version numbers in the assembly and NuGet manifest.

Well, then I suggest to remove `Publish Nuget packages to Nuget registry` step from ci workflow for now and merge as is (after your review) since proposed CI should be equal to all scripts before. You can publish releases the same way as you did before - manually. Further activity about publishing process (CD) will be tracked here - #258.


> Why aren't you a maintainer at this repo btw

No one invited me.

**Comment by stakx** [comment]
> No one invited me.

I see. Perhaps it wasn't clear in #210 that you were interested, too? If you still are, let me know.

>Someone writes PR. CI triggers, builds code, runs tests. No nuget publishing here, or you will end up with hundreds trash versions of package on nuget.org.

Oh, I think there's a misunderstanding. Of course publishing every single commit on `master` to NuGet would be overkill. I was only referring to a CI pipeline that publishes to NuGet; that one wouldn't have to publish the same NuGet package as a GitHub release artifact.

> we chose the second strategy for publishing to nuget.org but each commit to `master` or `develop` still results in publishing packages to `nuget.pkg.github.com`. Why? Because someone may have a strong need to obtain code-fixes (or test features) [...]

I see what you mean, makes sense. Not sure if it's really required for PublicApiGenerator which, I believe, has had relatively quick release cycles / iterations so users probably don't have to wait for very long for an official release (and they can always just as for one). It's not like we're planning several milestones far into the future, we're mostly doing small bugfixes or improvements that can be published more or less right away.

That being said, I have no problem with the release scheme you described.

It's getting late here, I'll need to postpone the review to tomorrow or the day after.

**Comment by sungam3r** [comment]
>  If you still are, let me know.

@stakx I could help to move things a bit faster here.

File: src/Directory.Build.props
Comment (stakx) [inline src/Directory.Build.props:11]: Making projects "packable" by default seems a little dangerous. Say that in the NuGet publishing step, we simply scan for all `*.nupkg` and publish them, and say that for some reason we add a second unit test project in the future but we forget to explicitly set its `<IsPackable>` property to `False`; we'd then publish the unit tests accidentally, wouldn't we?

I believe it would be safer to explicitly opt-in to `<IsPackable>`, rather than having to opt-out.

**Review by stakx — COMMENTED** [review]
Overall, these changes look plausible and sensible to me.

I haven't yet had the time to actually run a build with these changes, but am I understanding correctly that...

* You're removing `build.{sh,cmd}` because your MSBuild refactor means one can simply run `dotnet build`  directly instead?

* You're moving certain repeated project settings (such as the setup for strong-naming assemblies) into `Directory.Build.{props,targets}`?

**Comment by sungam3r** [comment]
> You're removing build.{sh,cmd} because your MSBuild refactor means one can simply run dotnet build directly instead?

Yes and no. These files IMO are needed only for CI on the remote machine. I "unwrapped" their contents into workflow. Of course developer **may** run such files locally as well but the need for this is extremely unlikely.

> You're moving certain repeated project settings (such as the setup for strong-naming assemblies) into Directory.Build.{props,targets}?

Yes. I am used to following this strategy in all the projects which I work on.

File: src/Directory.Build.props
Comment (sungam3r) [reply src/Directory.Build.props:11]: In fact, I spent a lot of time to comprehend what strategy for this property (in conjunction with others) is the most correct and universal for any .NET repo. There are several variations that one way or another lead to the desired result. For such a small project as this I can apply another option that you are talking about.

Comment (sungam3r) [reply src/Directory.Build.props]: done

**Comment by stakx** [comment]
I may still be missing something, but now that you've removed `<IsPackable>` everywhere (along with `<GeneratePaclageOnBuild>`) will `.nupkg`s still be produced where appropriate? Also it seems that since you've made strong-naming conditional (i.e. dependent on `<IsPackable>`) you essentially also stopped strong-naming the assemblies now...?

**Comment by sungam3r** [comment]
> I may still be missing something, but now that you've removed <IsPackable> everywhere (along with <GeneratePackageOnBuild>) will .nupkgs still be produced where appropriate?

Yes, because `IsPackable` by default is true for libraries and global tools. Also `IsPackable` is false for all test projects- i.e. projects that reference `Microsoft.NET.Test.Sdk` package. MS took care of a reasonable system of default values.

> Also it seems that since you've made strong-naming conditional (i.e. dependent on <IsPackable>) you essentially also stopped strong-naming the assemblies now...?

All test projects are not signed (as well as samples, dummy assemblies if any in the future). All other projects (packable ones) are signed.

**Comment by sungam3r** [comment]
@stakx OK to merge?

File: src/PublicApiGenerator.Tool/PublicApiGenerator.Tool.csproj
Comment (stakx) [inline src/PublicApiGenerator.Tool/PublicApiGenerator.Tool.csproj:5]: Just wondering whether or not we could pick better targets here. `netcoreapp2.1` is relatively ancient, do we still need that? (I don't mind but would probably have picked `netcoreapp3.1` instead.) And instead of `net7` I'd probably choose `net6.0` which is an LTS release, and should work for .NET 7, too.

**Comment by stakx** [comment]
@sungam3r, I suppose so. I'll leave this in your capable hands. If we notice any CI issues, we can always improve those incrementally.

**Review by stakx — APPROVED** [review]

**Comment by stakx** [comment]
Feel free to go ahead and merge.

Just one unrelated request: I notice that you disabled merge commits and rebases. Could we please undo this? I disagree with squash-merging only; the type of merge should be a case-by-case decision.

File: src/PublicApiGenerator.Tool/PublicApiGenerator.Tool.csproj
Comment (sungam3r) [reply src/PublicApiGenerator.Tool/PublicApiGenerator.Tool.csproj]: I'm fine either way. I changed to 3.1 + 6.

**Comment by sungam3r** [comment]
I did it so to keep history clean. If you look at https://github.com/PublicApiGenerator/PublicApiGenerator/commits/master then you will see commits with messages like `Merge pull request #xxx from yyy` or even worse:
изображение

These are temporary commits, which should not be in the master branch, they only clog history and do not provide a holistic picture of the work done, breaking it into many small pieces.

> the type of merge should be a case-by-case decision.

😕 I don't know. In practice, I have not seen cases when I had to choose between merge/rebase and squash.I returned merge/rebase options back but I will never use them for my own PRs.

**Comment by sungam3r** [comment]
@stakx Regarding TFMs:
изображение

IMO that was a reason why I chose 2.1 over 3.1.

**Pull request merged**
